### PR TITLE
Allow resolve to return non-existent resources

### LINKED ISSUE #11411: Allow non existant resources to be obtained from Resource.resolve()
**Jetty version(s)**
12

**Description**
Currently `Resource.resolve()` will return null if the resource does not exist.
This makes `Resource` harder to use (see case in PR #11279).

We have `Resource.exists()` method to determine whether a given resource exists, so why can't we get resources from `resolve` which do not exist?

The implementation of `PathResource.resolve()` does this:
```java
if (Files.exists(path))
    return newResource(path, resolvedUri);
return null;
```

### PR DESCRIPTION
**Pull request by gregw**
Fix #11411 by allowing non-existent resources to be returned from resolve

### CODE REVIEW

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/Resource.java [hunk 1]
```diff
@@ -268,7 +268,8 @@ public List<Resource> list()
      * Resolve an existing Resource.
      *
      * @param subUriPath the encoded subUriPath
-     * @return an existing Resource representing the requested subUriPath, or null if resource does not exist.
+     * @return an existing Resource representing the requested subUriPath, which may not {@link #exists() exist},
```
Comment (joakime) [inline jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/Resource.java]:
```suggestion
     * @return a Resource representing the requested subUriPath, which may not {@link #exists() exist},
```

**Comment by joakime** [comment]
Are we going to need to reintroduce a bad resource impl? something like EmptyResource? or InvalidResource? that specific ResourceFactory implementations can use when it's known that the requested resource (be it from `newResource()` or `.resolve()`) can never be valid? (eg from URL Resource, specific PathResource schemes, mounted Resources, etc)

Theoretically, this could also be used for exceptions/errors from the resource, so that .newResource() and .resolve() never throw (not sure how much benefit that would be though)

**Comment by gregw** [comment]
> Are we going to need to reintroduce a bad resource impl? something like EmptyResource?

I'd prefer not.

There vaste bulk of calls use Resources method to check the return, so null first impossible and non exists for possible is ok.

I'll try to make this work this way.

**Comment by gregw** [comment]
@janbartel

**Comment by gregw** [comment]
@lorban @joakime nudge

**Comment by lorban** [comment]
If my understanding is correct, this change would fix an API oddity:

`ResourceFactory.newResource("file:/this-directory-exists/this-file-does-not-exist")` returns a `PathResource` while `ResourceFactory.newResource("file:/this-directory-exists").resolve("this-file-does-not-exist")` returns null.

I think this is a good change, but that should be thoroughly reflected in our test suite as I'm sure methods like `Resource.copyTo(Path)` assume the resource exists.

(As a slightly unrelated note mostly to myself, `Resource` has many implementations expecting a `Path` that should probably be moved to `PathResource`, and I'm starting to think that `Resource.getPath()` was a mistake and should have been added only to `PathResource` and we should have used casts to `PathResource` where we know we can expect a `Resource` to be backed by a `Path`.)

**Comment by gregw** [comment]
@lorban I think the test suite is pretty good for resources.  I have also walked through all the calls to resolve and checked that they are using something like `Resources.missing(r)` or equivalent.  There are still a few null checks, but they are mostly in servlet land, where we do convert non existent resources into null.

Do you have anything specific you want tested?

@joakime nudge

**Comment by lorban** [comment]
@gregw `Resource.copyTo()` is poorly covered and has different behavior depending if the resource is backed by a path or not. Here is what I found so far:

 - If the resource is backed by a `Path` and is a regular file, `copyTo()` throws `UnsupportedOperationException: Unsupported copy option: ATOMIC_MOVE`.
 - If the resource is backed by a `Path` but does not exist, `copyTo()` trips on `assert isDirectory()`.
 - If the resource is a regular file:
   - If it is backed by a `Path`, `copyTo(directory)` creates or overwrites the file into the destination directory.
   - If it is NOT backed by a `Path`, `copyTo(file)` throws `FileSystemException: <target path>: Is a directory`.

I've already created https://github.com/jetty/jetty.project/pull/11493 regarding the 1st item, so maybe these extra tests and fixes should be taken care of in that PR too?

**Comment by gregw** [comment]
@lorban How are these `copyTo` issues related to this PR about `resolve` ?  Can't they all be addressed in #11493

**Comment by lorban** [comment]
This PR introduced this bug: _if the resource is backed by a Path but does not exist, copyTo() trips on assert isDirectory()_.

The others can be tackled in https://github.com/jetty/jetty.project/pull/11493.

**Review by lorban — APPROVED** [review]

**Comment by gregw** [comment]
@joakime I'd really like to get this merged soon, so it is not merged just before the next release.

**Review by lorban — CHANGES_REQUESTED** [review]
This needs to be fixed: if the resource is backed by a Path but does not exist, copyTo() trips on `assert isDirectory()`.

**Comment by joakime** [comment]
> This needs to be fixed: if the resource is backed by a Path but does not exist, copyTo() trips on `assert isDirectory()`.

Looks like you have that covered in the new testing in PR #11493
Wouldn't that fix be enough?

If we merge PR #11493 and then merge that into here, we should be able to validate this concern is fixed here too, right?

**Comment by lorban** [comment]
@joakime AFAICT it is not possible to create a `Resource` that does not exist without the changes of this PR, so this problem can only be triggered because of these changes. Or did I overlook something?

**Comment by joakime** [comment]
> @joakime AFAICT it is not possible to create a `Resource` that does not exist without the changes of this PR, so this problem can only be triggered because of these changes. Or did I overlook something?

This PR introduce the ability to create a `Resource` that doesn't yet exist.
You got it right.

**Comment by lorban** [comment]
@joakime thanks for the confirmation.
I pushed a fix for that bug as well as a new test for making sure all methods work on non-existent resources.

**Comment by gregw** [comment]
@lorban I think you used the wrong checkstyle setup as it still fails.  I have fixed.

**Review by lorban — APPROVED** [review]

**Pull request merged**